### ISSUE TITLE
globalScope feature feedback

### ISSUE THREAD

**Issue by ChristopherBiscardi** [ChristopherBiscardi/gatsby-mdx#226]
A couple ideas have surfaced in #138 so I'm creating a new issue to garner feedback around the new `globalScope` feature and how the API should evolve in the future.

example: A [components folder](https://github.com/ChristopherBiscardi/gatsby-mdx/issues/80) for gatsby-mdx.

**Comment by johno** [comment]
I kinda like the idea of a special mdx-components directory or file, tbh. It'd allow you to handle different import/export cases without having to learn new syntax or config keys.

```js
// ./src/components/_mdx-components.js
import { SketchPicker as Picker } from 'react-color'
import * as Foo from 'bar'
import MUIButton from '@material/core/button'

Foo.Baz = () => <h1>baz</h1>

export { Foo, MUIButton }
export { default as Button } from './src/component/button.js'
```

**Comment by jlengstorf** [comment]
I'm pretty into a special file, too. The tricky part, I guess, would be that if we actually ran it as code, we'd end up with something like this:

```
import * as GlobalScope from './src/componets/_mdx-components.js';

// How do we make `Picker` resolve from `GlobalScope.Picker`?
```

Maybe there's an easy way around that. A wrapper, maybe?

**Comment by jlengstorf** [comment]
Also we'd need to handle a `default` export from `_mdx-components.js`. Is that just not allowed (probably easiest), or is there special handling?

**Comment by ChristopherBiscardi** [comment]
If we went with a single file, I'd say don't use the default export and just use `export const` for all components, then we can automatically check for MDX components (`p`, etc) and put them in `MDXProvider` in the `wrapRootElement` and leave the rest for `MDXScopeProvider`... but we could easily check to see if the default import was an object and merge it in with named exports or whatever. It's a non-issue IMO, just requires some edge-case handling code.

**Comment by ChristopherBiscardi** [comment]
also, pointing to a file doesn't really mix well with themes without some additional code that checks all the theme directories for the special file (but only if they have gatsby-mdx defined, etc)

**Comment by jlengstorf** [comment]
Does that mean we'd have to do e.g.:
```js
import { SketchPicker as Picker } from 'react-color';

export Picker;
```
?

Thinking about themes, I keep coming back to the idea of making it composable for e.g. plugins.

Like, `gatsby-theme-blog` exports something like `PullQuote` from `node_modules/gatsby-theme-blog/components/_mdx-components.js`, which we don't want to lose.

So being able to say something like adding a path that defaults to `./src/components/_mdx-components.js` and auto-adds everything it exports (with overrides for collisions?) is super appealing to me.

Assuming that it would do something like this:

```
const finalGlobalScope = [
  ...globalScopeFromThemes,
  ...globalScopeFromPlugins, 
  ...globalScopeFromGatsbyConfig,
];
```

I assume that's easier said than done, but this would be an ideal outcome in my mind.

**Comment by rasovica** [comment]
I am having trouble getting this to work, I have version 0.3.3 of gatsby-mdx installed and the following in the config:
```   
     {
      resolve: `gatsby-mdx`,
      options: {
          globalScope: `
            import Layout from '${__dirname}/src/components/layout/layout.jsx';
            export default {Layout};
          `,
      }
```

The import resolves without error but it is not found inside of .mdx.

```
## Test

<div><Layout /></div>
```
I would really like this to get to work as it would remove at least 10 lines of code from each mdx.

Am I doing something wrong? I can get rid of this error
![error](https://screenshotscdn.firefoxusercontent.com/images/8a80f233-9594-47ab-a1d7-771aa0c2de6e.png)

**Comment by shripadk** [comment]
@rasovica You have to modify your gatsby-node.js file too and create a mdx-runtime.js file as explained in the [kitchen-sink](https://github.com/ChristopherBiscardi/gatsby-mdx/tree/master/examples/kitchen-sink) example. I modified it for my own project and this is what I have so far:

**1. gatsby-node.js**

```js
/**
 * Implement Gatsby's Node APIs in this file.
 *
 * See: https://www.gatsbyjs.org/docs/node-apis/
 */

const path = require('path');

exports.createPages = ({ graphql, actions }) => {
  const { createPage } = actions;
  return new Promise((resolve, reject) => {
    resolve(
      graphql(`
        allMdx {
          edges {
            node {
              id
              tableOfContents
              parent {
                ... on File {
                  absolutePath
                  name
                  sourceInstanceName
                }
              }
            }
          }
        }
      `).then(result => {
        if (result.errors) {
          console.log(result.errors); // eslint-disable-line no-console
          reject(result.errors);
        }

        result.data.allMdx.edges.forEach(({ node }) => {
          const actualPath = node.parent.absolutePath
            .replace('.mdx', '')
            .replace(`${__dirname}/src/pages`, '');

          const options = {
            component: path.resolve('./src/gatsby-components/mdx-runtime.js'),
            context: {
              absPath: node.parent.absolutePath,
              tableOfContents: node.tableOfContents,
              id: node.id,
            },
          };

          createPage({
            path: actualPath,
            ...options,
          });
        });
      }),
    );
  });
};

exports.onCreateWebpackConfig = ({ actions }) => {
  actions.setWebpackConfig({
    resolve: {
      modules: [path.resolve(__dirname, 'src'), 'node_modules'],
    },
  });
};
```

**2. mdx-runtime.js:**

```js
import React, { Component } from 'react';
import { graphql } from 'gatsby';

import MDXRenderer from 'gatsby-mdx/mdx-renderer';

// The layout that you were using earlier can be imported here
import { Layout } from './layout';

export default class MDXRuntime extends Component {
  render() {
    const { data, ...props } = this.props;

    return (
      <Layout {...props}>
        <MDXRenderer {...props}>{data.mdx.code.body}</MDXRenderer>
      </Layout>
    );
  }
}

export const pageQuery = graphql`
  query($id: String!) {
    mdx(id: { eq: $id }) {
      id
      code {
        body
      }
    }
  }
`;
```

The only issue currently with the `globalScope` feature is that you would need to restart the entire dev server for any changes/imports.

**Comment by ChristopherBiscardi** [comment]
@shripadk 

> You have to modify your gatsby-node.js file too and create a mdx-runtime.js file as explained in the kitchen-sink example.

You shouldn't have to do all that for `globalScope` to work.

@rasovica Do you mean that it isn't working for `.mdx` files in `src/pages`?

**Comment by rasovica** [comment]
@ChristopherBiscardi I was confused by @shripadk response as well because I followed your commit exactly.
Yes in a `.mdx` file in `src/pages`. I am pretty sure that I have a version of `gatsby-mdx` that has globalScope as I can console log from the code block and see that Layout component indeed gets imported

**Comment by shripadk** [comment]
@ChristopherBiscardi I tried it with a blank starter and it did not work as well. I'll put up the repo by end of the day for you to try.

@rasovica I followed the same guide as well. But it doesn't work and I got the same error as you did. So I followed the kitchen-sink example and realised that it only worked with modifications to gatsby-node.js and using a custom renderer.

This is my **gatsby-config.js** which you can verify for yourself:

```js
module.exports = {
  siteMetadata: {
    title: 'Example',
  },
  plugins: [
    'gatsby-plugin-react-helmet',
    'gatsby-plugin-styled-components',
    {
      resolve: `gatsby-plugin-manifest`,
      options: {
        name: 'gatsby-starter-default',
        short_name: 'starter',
        start_url: '/',
        background_color: '#663399',
        theme_color: '#663399',
        display: 'minimal-ui',
        icon: 'src/images/Icon.svg', // This path is relative to the root of the site.
      },
    },
    'gatsby-plugin-offline',
    'gatsby-plugin-typescript',
    {
      resolve: 'gatsby-mdx',
      options: {
        globalScope: `
          import { SketchPicker } from 'react-color';
  
          export default {
            SketchPicker,
          };
        `,
      },
    },
    {
      resolve: 'gatsby-source-filesystem',
      options: {
        path: `${__dirname}/src/pages`,
      },
    },
  ],
};

```

As you can see it does use `globalScope` as described in #138.

Can you try it with the above modifications and see if it works? If I comment **gatsby-node.js** code and use a default layout it stops working.

**Comment by shripadk** [comment]
@ChristopherBiscardi I have created a repo here: https://github.com/shripadk/gatsby-mdx-test for you to try out.

Just used the default gatsby-starter package with gatsby-mdx. The `master` branch is without the custom renderer and `with-runtime` branch is the one with the modifications. Please try it out and let me know if I am doing something wrong in the `master` branch.

The test page is located at `src/pages/test.mdx` and the location is http://localhost:8000

1. Testing from `master` branch:

<img width="3200" alt="screenshot 2019-01-09 at 12 38 29 pm" src="https://user-images.githubusercontent.com/189344/50882669-892af380-140b-11e9-986e-05a11f00eaf8.png">

2. Testing from `with-runtime` branch:

<img width="3200" alt="screenshot 2019-01-09 at 12 39 58 pm" src="https://user-images.githubusercontent.com/189344/50882718-b7a8ce80-140b-11e9-9c11-de2a9bf61ae1.png">

**Comment by ChristopherBiscardi** [comment]
@shripadk this issue is for feature API design feedback, not bug reports. Please open a new issue next time. Here's one for the issue you're both seeing right now: https://github.com/ChristopherBiscardi/gatsby-mdx/issues/239

**Comment by shripadk** [comment]
@ChristopherBiscardi I understand the difference. I wasn't aware that it was a bug in the first place. I thought that globalScope can only be used with modifications to gatsby-node.js and gatsby-config.js because I followed the kitchen-sink example. Only after your response did I realise that it was intended to be used without those modifications. So created and shared a repo to know if I was making any mistake in the config. Now that you say it's a bug i'll continue the discussion there. You can blame it on my excitement to try out the new API before it was stable 😄Sorry about that!

**Comment by alexandernanberg** [comment]
@ChristopherBiscardi Would it work if we would just context for this? So you would do `<MDXScopeProvider scope={components} />` in a layout/app component. Not a huge fan personally of the separate file or the `globalScope` options 🤔

**Comment by ChristopherBiscardi** [comment]
@alexandernanberg I recently submitted a PR to MDX to support this functionality through an enhanced MDXProvider. It landed in the [MDX 1.0 branch](https://github.com/mdx-js/mdx/tree/v1) so will be published then and gatsby-mdx@1.0 will remove the `globalScope` option because it can be served by using `MDXProvider` in `wrapRootElement`.

**Comment by alexandernanberg** [comment]
Ah that’s fantastic, looking forward to try it out 😄

**Comment by rwieruch** [comment]
Difficult to find the latest material on "how to define global scoped components with Gatsby MDX" :) The final culprit was not using `@mdx-js/react` but still `@mdx-js/tag`.

If anyone is interested in seeing global components in action: https://github.com/rwieruch/gatsby-mdx-blog-starter-project

Thanks for your hard work @ChristopherBiscardi ❤️

**Comment by eyalroth** [comment]
@ChristopherBiscardi I'm finding it impossible to make "shortcodes" work on my feed (`gatsby-plugin-feed`). I tried using an `MDXProvider` in both `wrapRootElement` methods (SSR and browser), plus importing my component via the `globalScope` option, and still the component is left unresolved in my feed HTML (it renders as `div`). I'm also seeing these messages when building:
```
Component X was not imported, exported, or provided by MDXProvider as global scope
```
My environment (WSL):
```
System:
    OS: Linux 4.4 Ubuntu 16.04.6 LTS (Xenial Xerus)
    CPU: (4) x64 Intel(R) Core(TM) i7-5600U CPU @ 2.60GHz
    Shell: 4.3.48 - /bin/bash
  Binaries:
    Node: 10.16.0 - ~/n/bin/node
    Yarn: 1.17.3 - /usr/bin/yarn
    npm: 6.11.3 - ~/n/bin/npm
  Languages:
    Python: 2.7.12 - /usr/bin/python
  npmPackages:
    gatsby: ^2.13.62 => 2.13.62
    gatsby-image: ^2.2.8 => 2.2.8
    gatsby-plugin-catch-links: ^2.1.3 => 2.1.3
    gatsby-plugin-feed: ^2.3.6 => 2.3.6
    gatsby-plugin-google-analytics: ^2.1.7 => 2.1.7
    gatsby-plugin-layout: ^1.1.3 => 1.1.3
    gatsby-plugin-manifest: ^2.2.5 => 2.2.5
    gatsby-plugin-mdx: ^1.0.43 => 1.0.43
    gatsby-plugin-netlify: ^2.1.4 => 2.1.4
    gatsby-plugin-offline: ^2.2.6 => 2.2.6
    gatsby-plugin-prefetch-google-fonts: ^1.4.3 => 1.4.3
    gatsby-plugin-react-helmet: ^3.1.3 => 3.1.3
    gatsby-plugin-robots-txt: ^1.5.0 => 1.5.0
    gatsby-plugin-sass: ^2.1.8 => 2.1.8
    gatsby-plugin-sharp: ^2.2.11 => 2.2.11
    gatsby-plugin-sitemap: ^2.2.6 => 2.2.6
    gatsby-remark-copy-linked-files: ^2.1.5 => 2.1.5
    gatsby-remark-images: ^2.0.4 => 2.0.6
    gatsby-remark-prismjs: ^3.3.5 => 3.3.5
    gatsby-remark-prismjs-title: ^1.0.0 => 1.0.0
    gatsby-remark-responsive-iframe: ^2.2.4 => 2.2.4
    gatsby-remark-smartypants: ^2.1.2 => 2.1.2
    gatsby-source-filesystem: ^2.1.9 => 2.1.9
    gatsby-transformer-remark: ^2.6.14 => 2.6.14
    gatsby-transformer-sharp: ^2.2.6 => 2.2.6
  npmGlobalPackages:
    gatsby-cli: 2.7.44
```
